### PR TITLE
Profile: allocate buffer for n instruction pointers per thread

### PR DESCRIPTION
**Pull request by IanButterworth**
Given that profiling samples all threads (except windows that only profiles the main thread) it seems to make sense to allocate nsamples per thread (except windows).

### CODE REVIEW

**Review by timholy — APPROVED** [review]

**Review by vchuravy — APPROVED** [review]

File: stdlib/Profile/src/Profile.jl [hunk 1]
```diff
@@ -40,7 +40,7 @@ end
     init(; n::Integer, delay::Real))
 
 Configure the `delay` between backtraces (measured in seconds), and the number `n` of
-instruction pointers that may be stored. Each instruction pointer corresponds to a single
+instruction pointers that may be stored per thread. Each instruction pointer corresponds to a single
```
Comment (vchuravy) [inline stdlib/Profile/src/Profile.jl:43]: Maybe add a `!!! compat` note?

Comment (NHDaly) [reply stdlib/Profile/src/Profile.jl:43]: +1 to a compat note

**Review by NHDaly — APPROVED** [review]

**Comment by vtjnash** [comment]
It looks like this broke the linux32 build, as seen in the pre-merge CI

**Comment by IanButterworth** [comment]
Sorry about that.

The only way I can think that this PR is causing the error below is if somehow the `nthreads` var inside `Profile.init()` is messing with `Threads.nthreads`... but.... .....
```
Error in testset cmdlineargs:
Error During Test at /buildworker/worker/tester_linux32/build/share/julia/test/cmdlineargs.jl:202
  Test threw exception
  Expression: string(cpu_threads) == read(`$exename --threads auto -e $code`, String) == read(`$exename --threads=auto -e $code`, String) == read(`$exename -tauto -e $code`, String) == read(`$exename -t auto -e $code`, String)
  failed process: Process(`/buildworker/worker/tester_linux32/build/bin/julia -Cnative -J/buildworker/worker/tester_linux32/build/lib/julia/sys.so --depwarn=error --check-bounds=yes -g1 --startup-file=no --startup-file=no --color=no --threads auto -e 'print(Threads.nthreads())'`, ProcessExited(1)) [1]
  
  Stacktrace:
   [1] pipeline_error
     @ ./process.jl:532 [inlined]
   [2] read(cmd::Cmd)
     @ Base ./process.jl:419
   [3] read(cmd::Cmd, #unused#::Type{String})
     @ Base ./process.jl:428
   [4] macro expansion
     @ /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.8/Test/src/Test.jl:445 [inlined]
   [5] top-level scope
     @ /buildworker/worker/tester_linux32/build/share/julia/test/cmdlineargs.jl:202
```

**Comment by vtjnash** [comment]
It is attempting to initialize profile space for 128 threads, and now failing (OOM)

**Comment by IanButterworth** [comment]
128 real threads, or is that a bug also? 

If that's real, then options seem:

1. Revert. It's not critical, more of a design tweak
2. Reduce the default buffer size (per thread). How much ram does the linux32 128 thread machine have? Perhaps that can guide this.

I guess it's not reliable that a n-thread machine would have n-times more ram than a 1-thread machine, which this PR kind of assumes

File: stdlib/Profile/src/Profile.jl [hunk 2]
```diff
@@ -40,26 +40,32 @@ end
...
 """
 function init(; n::Union{Nothing,Integer} = nothing, delay::Union{Nothing,Real} = nothing)
     n_cur = ccall(:jl_profile_maxlen_data, Csize_t, ())
     delay_cur = ccall(:jl_profile_delay_nsec, UInt64, ())/10^9
     if n === nothing && delay === nothing
         return Int(n_cur), delay_cur
     end
-    nnew = (n === nothing) ? n_cur : n
+    nthreads = Sys.iswindows() ? 1 : Threads.nthreads() # windows only profiles the main thread
+    nnew = (n === nothing) ? n_cur : n * nthreads
```
Comment (IanButterworth) [inline stdlib/Profile/src/Profile.jl:59]: There's also a bug here. This method passes to the method below, so this was multiplying by nthreads twice

**Pull request merged**